File: client_packages/RomaniaRoleplay/Characters/character-helper.js
var cameraCharacterPreview = null;
var defaultMaleModel = 1885233650;
var defaultFemaleModel = 2627665880;


var defaultSkin = {
    Model: 1885233650,
    FirstHeadShape: 0,
    SecondHeadShape: 0,
    FirstSkinTone: 0,
    SecondSkinTone: 0,
    HeadMix: -1,
    SkinMix: -1,
    Hair: 0,
    FirstHairColor: 0,
    SecondHairColor: 0,
    Beard: 0,
    BeardColor: 0,
    Chest: 0,
    ChestColor: 0,
    Blemishes: 0,
    Ageing: 0,
    Complexion: 0,
    Sundamage: 0,
    Freckles: 0,
    EyesColor: 0,
    Eyebrows: 0,
    EyebrowsColor: 0,
    Makeup: 0,
    Blush: 0,
    BlushColor: 0,
    Lipstick: 0,
    LipstickColor: 0,
    NoseWidth: -1.0,
    NoseHeight: -1.0,
    NoseLength: -1.0,
    NoseBridge: -1.0,
    NoseTip: -1.0,
    NoseShift: -1.0,
    BrowHeight: -1.0,
    BrowWidth: -1.0,
    CheekboneHeight: -1.0,
    CheekboneWidth: -1.0,
    CheeksWidth: -1.0,
    Eyes: -1.0,
    Lips: -1.0,
    JawWidth: -1.0,
    JawHeight: -1.0,
    ChinLength: -1.0,
    ChinPosition: -1.0,
    ChinWidth: -1.0,
    ChinShape: -1.0,
    NeckWidth: -1.0,
    Clothes: {
        Torsos: 0,
        TorsosColor: 0,
        Legs: 0,
        LegsColor: 0,
        BagsAndParachutes: 0,
        BagsAndParachutesColor: 0,
        Shoes: 0,
        ShoesColor: 0,
        Accessories: 0,
        AccessoriesColor: 0,
        Undershirts: 0,
        UndershirtsColor: 0,
        BodyArmors: 0,
        BodyArmorsColor: 0,
        Decals: 0,
        DecalsColor: 0,
        Tops: 0,
        TopsColor: 0
    }
};

var showCharacterPreview = function (shouldShow) {
    if (shouldShow) {
        try {
            mp.game.ui.displayRadar(false);
            cameraCharacterPreview = mp.cameras.new("Cam", {
                x: -422.38,
                y: 1134.75,
                z: 326.674
            }, {
                x: 0,
                y: 0,
                z: 163
            }, 6);
            cameraCharacterPreview.setActive(true);
            mp.game.cam.renderScriptCams(true, true, 20000000000000000000000000, false, false);
            player.position = new mp.Vector3(-427.65, 1116.374, 326.8);
            player.heading = 343;
            setTimeout(function () {
                mp.gui.cursor.show(true, true);
            }, 10);
            setTimeout(function () {
                player.freezePosition(true);
            }, 10);
            setTimeout(function () {
                mp.game.player.setInvincible(true);
            }, 10);
            setTimeout(function () {
                activateChat(false);
            }, 1000);
        } catch (e) {
            console.log(e);
        }
    } else {
        try {
            mp.game.ui.displayRadar(true);
            activateChat(true);
            cameraCharacterPreview.setActive(false);
            mp.game.cam.renderScriptCams(false, true, 0, true, true);
            cameraCharacterPreview.destroy();
            setTimeout(function () {
                mp.gui.cursor.show(false, false);
            }, 10);
            setTimeout(function () {
                player.freezePosition(false);
            }, 10);
            setTimeout(function () {
                mp.game.player.setInvincible(false);
            }, 10);
            setTimeout(function () {
                activateChat(true);
            }, 1000);
        } catch (e) {
            console.log(e);
        }
    }
};

var setCharacterLocal = function (character) {
    player.setMoney(character.Money);
    mp.game.stats.statSetInt(mp.game.joaat('SP0_TOTAL_CASH'), character.Money, true);
}

var HeadOverlayIds = {
    Blemishes: 0,
    FacialHair: 1,
    Eyebrows: 2,
    Ageing: 3,
    Makeup: 4,
    Blush: 5,
    Complexion: 6,
    SunDamage: 7,
    Lipstick: 8,
    Freckles: 9,
    ChestHair: 10,
    BodyBlemishes: 11,
    AddBodyBlemishes: 12
};

var ClothesIndex = {
    Masks: 1,
    Hair: 2,
    Torsos: 3,
    Legs: 4,
    BagsAndParachutes: 5,
    Shoes: 6,
    Accessories: 7,
    Undershirts: 8,
    BodyArmors: 9,
    Decals: 10,
    Tops: 11
}

var FaceFeatureIndexes = {
    NoseWidth: 0,
    NoseHeight: 1,
    NoseLength: 2,
    NoseBridge: 3,
    NoseTip: 4,
    NoseShift: 5,
    BrowHeight: 6,
    BrowWidth: 7,
    CheekboneHeight: 8,
    CheekboneWidth: 9,
    CheeksWidth: 10,
    Eyes: 11,
    Lips: 12,
    JawWidth: 13,
    JawHeight: 14,
    ChinLength: 15,
    ChinPosition: 16,
    ChinWidth: 17,
    ChinShape: 18,
    NeckWidth: 19,
}

var setCharacterSkin = function (skin) {
    if (skin === null || skin === undefined) {
        skin = defaultSkin;
    }

    player.model = skin.Model;

    player.setHeadBlendData(
        skin.FirstHeadShape,
        skin.SecondHeadShape,
        0,
        skin.FirstSkinTone,
        skin.SecondSkinTone,
        0,
        skin.HeadMix,
        skin.SkinMix,
        0,
        false);
    player.setComponentVariation(ClothesIndex.Hair, skin.Hair, 0, 2);
    player.setHairColor(skin.FirstHairColor, skin.SecondHairColor);
    player.setEyeColor(skin.EyesColor);

    player.setHeadOverlay(HeadOverlayIds.Blemishes, skin.Blemishes, 1.0, 0, 0);
    if (skin.Model == defaultMaleModel) {
        player.setHeadOverlay(HeadOverlayIds.FacialHair, skin.Beard, 1.0, skin.BeardColor, 0);
    } else {
        player.setHeadOverlay(HeadOverlayIds.Makeup, skin.Makeup, 1.0, 0, 0);
        player.setHeadOverlay(HeadOverlayIds.Blush, skin.Blush, 1.0, skin.BlushColor, 0);
        player.setHeadOverlay(HeadOverlayIds.Lipstick, skin.Lipstick, 1.0, skin.LipstickColor, 0);
    }
    player.setHeadOverlay(HeadOverlayIds.Eyebrows, skin.Eyebrows, 1.0, skin.EyebrowsColor, 0);
    player.setHeadOverlay(HeadOverlayIds.Ageing, skin.Ageing, 1.0, 0, 0);

    player.setHeadOverlay(HeadOverlayIds.Complexion, skin.Complexion, 1.0, 0, 0);
    player.setHeadOverlay(HeadOverlayIds.SunDamage, skin.Sundamage, 1.0, 0, 0);
    player.setHeadOverlay(HeadOverlayIds.Freckles, skin.Freckles, 1.0, 0, 0);
    player.setHeadOverlay(HeadOverlayIds.ChestHair, skin.Chest, 1.0, skin.ChestColor, 0);
    //player.setHeadOverlay(HeadOverlayIds.BodyBlemishes, skinShowedDefault.BodyBlemishes, 1.0, skinShowedDefault.FirstHairColor, skinShowedDefault.SecondHairColor);
    //player.setHeadOverlay(HeadOverlayIds.AddBodyBlemishes, skinShowedDefault.AddBodyBlemishes, 1.0, skinShowedDefault.FirstHairColor, skinShowedDefault.SecondHairColor);

    player.setFaceFeature(FaceFeatureIndexes.NoseWidth, skin.NoseWidth);
    player.setFaceFeature(FaceFeatureIndexes.NoseHeight, skin.NoseHeight);
    player.setFaceFeature(FaceFeatureIndexes.NoseLength, skin.NoseLength);
    player.setFaceFeature(FaceFeatureIndexes.NoseBridge, skin.NoseBridge);
    player.setFaceFeature(FaceFeatureIndexes.NoseTip, skin.NoseTip);
    player.setFaceFeature(FaceFeatureIndexes.NoseShift, skin.NoseShift);
    player.setFaceFeature(FaceFeatureIndexes.BrowHeight, skin.BrowHeight);
    player.setFaceFeature(FaceFeatureIndexes.BrowWidth, skin.BrowWidth);
    player.setFaceFeature(FaceFeatureIndexes.CheekboneHeight, skin.CheekboneHeight);
    player.setFaceFeature(FaceFeatureIndexes.CheekboneWidth, skin.CheekboneWidth);
    player.setFaceFeature(FaceFeatureIndexes.CheeksWidth, skin.CheeksWidth);
    player.setFaceFeature(FaceFeatureIndexes.Eyes, skin.Eyes);
    player.setFaceFeature(FaceFeatureIndexes.Lips, skin.Lips);
    player.setFaceFeature(FaceFeatureIndexes.JawWidth, skin.JawWidth);
    player.setFaceFeature(FaceFeatureIndexes.JawHeight, skin.JawHeight);
    player.setFaceFeature(FaceFeatureIndexes.ChinLength, skin.ChinLength);
    player.setFaceFeature(FaceFeatureIndexes.ChinPosition, skin.ChinPosition);
    player.setFaceFeature(FaceFeatureIndexes.ChinWidth, skin.ChinWidth);
    player.setFaceFeature(FaceFeatureIndexes.ChinShape, skin.ChinShape);
    player.setFaceFeature(FaceFeatureIndexes.NeckWidth, skin.NeckWidth);

    player.setComponentVariation(ClothesIndex.Torsos, skin.Clothes.Torsos, skin.Clothes.TorsosColor, 2);
    player.setComponentVariation(ClothesIndex.Legs, skin.Clothes.Legs, skin.Clothes.LegsColor, 2);
    player.setComponentVariation(ClothesIndex.BagsAndParachutes, skin.Clothes.BagsAndParachutes, skin.Clothes.BagsAndParachutesColor, 2);
    player.setComponentVariation(ClothesIndex.Shoes, skin.Clothes.Shoes, skin.Clothes.ShoesColor, 2);
    player.setComponentVariation(ClothesIndex.Accessories, skin.Clothes.Accessories, skin.Clothes.AccessoriesColor, 2);
    player.setComponentVariation(ClothesIndex.Undershirts, skin.Clothes.Undershirts, skin.Clothes.UndershirtsColor, 2);
    player.setComponentVariation(ClothesIndex.BodyArmors, skin.Clothes.BodyArmors, skin.Clothes.BodyArmorsColor, 2);
    player.setComponentVariation(ClothesIndex.Decals, skin.Clothes.Decals, skin.Clothes.DecalsColor, 2);
    player.setComponentVariation(ClothesIndex.Tops, skin.Clothes.Tops, skin.Clothes.TopsColor, 2);
};

File: dotnet/resources/RomaniaRoleplay/RomaniaRoleplay/Helpers/CharacterHelper.cs
using Domain.Characters.Models;
using GTANetworkAPI;
using RomaniaRoleplay.Helpers.Enums;
using System;
using System.Collections.Generic;

namespace RomaniaRoleplay.Helpers
{
    internal static class ClientCharacterHelper
    {
        public static void SetPlayerCharacter(Player player, Character character)
        {
            NAPI.Player.SetPlayerName(player, character.Name);
            NAPI.Player.SetPlayerNametag(player, $"{character.NameTag}(${player.Id})");

            SetCharacterCustomization(player, character);
            SetCharacterClothes(player, character);
        }

        public static void SetCharacterClothes(Player player, Character character, ClothesIndex? specificClotheIndex = null)
        {
            Dictionary<int, ComponentVariation> clothes = new Dictionary<int, ComponentVariation>();
            if (specificClotheIndex.HasValue)
            {
                AddClothesIndexToDictionary(specificClotheIndex.Value, clothes, character);
            }
            else
            {
                foreach (ClothesIndex overlay in Enum.GetValues(typeof(ClothesIndex)))
                {
                    AddClothesIndexToDictionary(overlay, clothes, character);
                }
            }
            player.SetClothes(clothes);
        }

        public static void SetCharacterCustomization(Player player, Character character)
        {
            HeadBlend headBlend = SetHeadBlend(character.Skin);

            var eyeColor = Convert.ToByte(character.Skin.EyesColor);
            var hairColor = Convert.ToByte(character.Skin.FirstHairColor);
            var highlightColor = Convert.ToByte(character.Skin.SecondHairColor);

            float[] faceFeatures = SetFaceFeatures(character.Skin);

            Dictionary<int, HeadOverlay> headOverlays = SetHeadOverlay(character.Skin);

            bool isMale = character.Skin.Model == (uint)PedHash.FreemodeMale01;

            player.SetCustomization(isMale, headBlend, eyeColor, hairColor, highlightColor, faceFeatures, headOverlays, new Decoration[] { });

        }

        private static void AddClothesIndexToDictionary(ClothesIndex clothesIndex, Dictionary<int, ComponentVariation> clothes, Character character)
        {
            switch (clothesIndex)
            {
                case ClothesIndex.Masks:
                    clothes.TryAdd((int)ClothesIndex.Masks, new ComponentVariation(0, 0));
                    break;
                case ClothesIndex.Hair:
                    clothes.TryAdd((int)ClothesIndex.Hair, new ComponentVariation(character.Skin.Hair, 0));
                    break;
                case ClothesIndex.Torsos:
                    clothes.TryAdd((int)ClothesIndex.Torsos, new ComponentVariation(character.Skin.Clothes.Torsos, 0));
                    break;
                case ClothesIndex.Legs:
                    clothes.TryAdd((int)ClothesIndex.Legs, new ComponentVariation(character.Skin.Clothes.Legs, 0));
                    break;
                case ClothesIndex.BagsAndParachutes:
                    clothes.TryAdd((int)ClothesIndex.BagsAndParachutes, new ComponentVariation(character.Skin.Clothes.BagsAndParachutes, 0));
                    break;
                case ClothesIndex.Shoes:
                    clothes.TryAdd((int)ClothesIndex.Shoes, new ComponentVariation(character.Skin.Clothes.Shoes, 0));
                    break;
                case ClothesIndex.Accessories:
                    clothes.TryAdd((int)ClothesIndex.Accessories, new ComponentVariation(character.Skin.Clothes.Accessories, 0));
                    break;
                case ClothesIndex.Undershirts:
                    clothes.TryAdd((int)ClothesIndex.Undershirts, new ComponentVariation(character.Skin.Clothes.Undershirts, 0));
                    break;
                case ClothesIndex.BodyArmors:
                    clothes.TryAdd((int)ClothesIndex.BodyArmors, new ComponentVariation(character.Skin.Clothes.BodyArmors, 0));
                    break;
                case ClothesIndex.Decals:
                    clothes.TryAdd((int)ClothesIndex.Decals, new ComponentVariation(character.Skin.Clothes.Decals, 0));
                    break;
                case ClothesIndex.Tops:
                    clothes.TryAdd((int)ClothesIndex.Tops, new ComponentVariation(character.Skin.Clothes.Tops, 0));
                    break;
            }
        }

        private static HeadBlend SetHeadBlend(Skin skin)
        {
            HeadBlend headBlend = new HeadBlend
            {
                ShapeFirst = Convert.ToByte(skin.FirstHeadShape),
                ShapeSecond = Convert.ToByte(skin.SecondHeadShape),
                ShapeMix = skin.HeadMix,
                SkinFirst = Convert.ToByte(skin.FirstSkinTone),
                SkinSecond = Convert.ToByte(skin.SecondHeadShape),
                SkinMix = skin.SkinMix
            };
            return headBlend;
        }

        private static float[] SetFaceFeatures(Skin skin)
        {
            float[] faceFeatures = new float[]
            {
                skin.NoseWidth, skin.NoseHeight, skin.NoseLength, skin.NoseBridge, skin.NoseTip, skin.NoseShift, skin.BrowHeight,
                skin.BrowWidth, skin.CheekboneHeight, skin.CheekboneWidth, skin.CheeksWidth, skin.Eyes, skin.Lips, skin.JawWidth,
                skin.JawHeight, skin.ChinLength, skin.ChinPosition, skin.ChinWidth, skin.ChinShape, skin.NeckWidth
            };
            return faceFeatures;
        }

        private static Dictionary<int, HeadOverlay> SetHeadOverlay(Skin skin)
        {
            Dictionary<int, HeadOverlay> headOverlays = new Dictionary<int, HeadOverlay>();

            foreach (HeadOverlayIndexes overlay in Enum.GetValues(typeof(HeadOverlayIndexes)))
            {
                if (skin.Model == (uint)PedHash.FreemodeMale01 && (overlay == HeadOverlayIndexes.Lipstick || overlay == HeadOverlayIndexes.Blush || overlay == HeadOverlayIndexes.Makeup))
                {
                    goto SkipCustomization;
                }
                if (skin.Model == (uint)PedHash.FreemodeFemale01 && overlay == HeadOverlayIndexes.FacialHair)
                {
                    goto SkipCustomization;
                }
                // Get the overlay model and color
                (int index, int color) = GetOverlayData(skin, overlay);

                // Create the overlay
                HeadOverlay headOverlay = new HeadOverlay();
                headOverlay.Index = Convert.ToByte(index);
                headOverlay.Color = Convert.ToByte(color);
                headOverlay.SecondaryColor = 0;
                headOverlay.Opacity = 1.0f;

                // Add the overlay
                headOverlays.TryAdd((int)overlay, headOverlay);

            SkipCustomization: continue;
            }
            return headOverlays;
        }

        private static (int index, int color) GetOverlayData(Skin skin, HeadOverlayIndexes overlay)
        {
            switch (overlay)
            {
                case HeadOverlayIndexes.Blemishes:
                    return (skin.Blemishes, 0);
                case HeadOverlayIndexes.FacialHair:
                    return (skin.Beard, skin.BeardColor);
                case HeadOverlayIndexes.Eyebrows:
                    return (skin.Eyebrows, skin.EyebrowsColor);
                case HeadOverlayIndexes.Ageing:
                    return (skin.Ageing, 0);
                case HeadOverlayIndexes.Makeup:
                    return (skin.Makeup, 0);
                case HeadOverlayIndexes.Blush:
                    return (skin.Blush, skin.BlushColor);
                case HeadOverlayIndexes.Complexion:
                    return (skin.Complexion, 0);
                case HeadOverlayIndexes.SunDamage:
                    return (skin.Sundamage, 0);
                case HeadOverlayIndexes.Lipstick:
                    return (skin.Lipstick, skin.LipstickColor);
                case HeadOverlayIndexes.Freckles:
                    return (skin.Freckles, 0);
                case HeadOverlayIndexes.ChestHair:
                    return (skin.Chest, skin.ChestColor);
                default:
                    return (0, 0);
            }
        }
    }
}
